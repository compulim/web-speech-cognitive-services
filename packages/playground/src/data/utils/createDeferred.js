// The MIT License (MIT)
//
// Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (sindresorhus.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// This piece of code is adopted from https://github.com/sindresorhus/p-defer

// The reason why we need to fork it is because:
// - The original package published to NPM is not ES5-compliant
//    - Due to the use of arrow functions
// - create-react-app@1 does not play nice with packages that are not ES5-compliant
//    - create-react-app@2 do play nice, but it was so new that most of the people are still on @1

// Criteria to remove this package:
// - When create-react-app@2 become mainstream

module.exports = function() {
  const ret = {};

  ret.promise = new Promise(function(resolve, reject) {
    ret.resolve = resolve;
    ret.reject = reject;
  });

  return ret;
};
