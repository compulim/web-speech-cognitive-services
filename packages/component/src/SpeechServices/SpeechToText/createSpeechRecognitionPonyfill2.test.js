jest.useFakeTimers();

import { PromiseHelper } from 'microsoft-cognitiveservices-speech-sdk/distrib/lib/src/common/Promise';
import createDeferred from '../../Util/createDeferred';

const MOCK_SPEECH_SDK = {
  AudioConfig: {
    fromDefaultMicrophoneInput: () => {
      const eventHandlers = [];
      const readResolves = [];

      return ({
        attach: () => PromiseHelper.fromResult({
          read: () => ({
            onSuccessContinueWith: resolve => readResolves.push(resolve)
          })
        }),
        emitEvent: name => {
          eventHandlers.forEach(handler => handler({ name }));
        },
        emitRead: chunk => {
          readResolves.forEach(resolve => resolve(chunk));
          readResolves.splice(0);
        },
        events: {
          attach: handler => {
            eventHandlers.push(handler);

            return {
              detach: () => eventHandlers.splice(eventHandlers.indexOf(handler), 1)
            };
          }
        }
      });
    }
  },
  OutputFormat: {
    Detailed: 'DETAILED'
  },
  ResultReason: {
    NoMatch: 0,
    Canceled: 1,
    RecognizingSpeech: 2,
    RecognizedSpeech: 3
  },
  SpeechConfig: {
    fromSubscription: (subscriptionKey, region) => {
      return {
        region,
        subscriptionKey
      };
    }
  },
  SpeechRecognizer: class {
    constructor(speechConfig, audioConfig) {
      this.audioConfig = audioConfig;
      this.speechConfig = speechConfig;
      this.canceled = this.recognized = this.recognizing = () => {};

      this.callRecognizeOnceAsyncDeferred = createDeferred();
      this.callStopContinuousRecognitionAsyncDeferred = createDeferred();
    }

    dispose() {}

    recognizeOnceAsync(...args) {
      this.callRecognizeOnceAsyncDeferred.resolve(args);
    }

    stopContinuousRecognitionAsync(...args) {
      this.callStopContinuousRecognitionAsyncDeferred.resolve(args);
    }

    async setFirstAudioChunk(chunk = []) {
      this.audioConfig.attach().onSuccessContinueWith(reader => reader.read(chunk));
    }

    async waitForRecognizeOnceAsync() {
      return await this.callRecognizeOnceAsyncDeferred.promise;
    }

    async waitForStopContinuousRecognitionAsync() {
      return await this.callStopContinuousRecognitionAsyncDeferred.promise;
    }

    async rejectRecognizeOnceAsync(err) {
      return (await this.callRecognizeOnceAsyncDeferred.promise)[1](err);
    }

    async resolveRecognizeOnceAsync(result) {
      return (await this.callRecognizeOnceAsyncDeferred.promise)[0](result);
    }
  }
};

const MOCK_RECOGNIZING_EVENT = {
  result: {
    duration: 1,
    json: JSON.stringify({
      Duration: 1,
      Offset: 0,
      Text: 'hello'
    }),
    offset: 0,
    reason: 2,
    text: 'hello'
  }
};

const MOCK_RECOGNIZED_EVENT = {
  result: {
    duration: 2,
    json: JSON.stringify({
      Duration: 2,
      Offset: 0,
      NBest: [{
        Confidence: 0.9,
        Lexical: 'hello john',
        ITN: 'hello John',
        MaskedITN: 'hello John',
        Display: 'Hello, John.'
      }]
    }),
    offset: 0,
    reason: 3,
    text: 'Hello, John.'
  }
};

const MOCK_SUCCESS_RESULT = {
  duration: 2,
  json: JSON.stringify({
    Duration: 2,
    Offset: 0,
    NBest: [{
      Confidence: 0.9,
      Lexical: 'hello john',
      ITN: 'hello John',
      MaskedITN: 'hello John',
      Display: 'Hello, John.'
    }]
  }),
  offset: 0,
  reason: 3,
  text: 'Hello, John.'
};

const SPEECH_EVENTS = [
  'audioend',
  'audiostart',
  'end',
  'error',
  'result',
  'soundend',
  'soundstart',
  'speechend',
  'speechstart',
  'start'
];

function captureSpeechEvents(speech) {
  const queue = [];

  for (let eventName of SPEECH_EVENTS) {
    switch (eventName) {
      case 'error':
        speech.addEventListener(eventName, ({ error }) => queue.push(`webspeech:error { error: '${ error }' }`));

        break;

      case 'result':
        speech.addEventListener(eventName, ({ results, type }) => {
          const { length } = results;
          const [{ isFinal } = {}] = results;

          queue.push(`webspeech:${ type } { isFinal: ${ !!isFinal }, length: ${ length } }`);
        });

        break;

      default:
        speech.addEventListener(eventName, ({ type }) => queue.push(`webspeech:${ type }`));

        break;
    }
  }

  speech.addEventListener('cognitiveservices', ({ subType, type }) => queue.push(`${ type }:${ subType }`));

  return queue;
}

describe('Mock SpeechRecognizer with', () => {
  let endEventEmitted;
  let errorEventEmitted;
  let events;
  let recognizer;
  let speechRecognition;

  beforeEach(async () => {
    jest.resetModules();
    jest.setMock('../SpeechSDK', ({
      ...MOCK_SPEECH_SDK,
      SpeechRecognizer: class extends MOCK_SPEECH_SDK.SpeechRecognizer {
        constructor(...args) {
          super(...args);
          recognizer = this;
        }
      }
    }));

    global.window = {
      document: {
        documentElement: {
          getAttribute: attributeName => {
            if (attributeName === 'lang') {
              return 'en-US';
            } else {
              throw new Error('not implemented');
            }
          }
        }
      },
      navigator: {
        mediaDevices: {
          getUserMedia: () => {}
        }
      }
    };

    const { default: createSpeechRecognitionPonyfill } = require('./createSpeechRecognitionPonyfill');
    const { SpeechRecognition } = await createSpeechRecognitionPonyfill({
      region: 'westus',
      subscriptionKey: 'SUBSCRIPTION_KEY'
    });

    speechRecognition = new SpeechRecognition();
    events = captureSpeechEvents(speechRecognition);
    endEventEmitted = new Promise(resolve => speechRecognition.addEventListener('end', resolve));
    errorEventEmitted = new Promise(resolve => speechRecognition.addEventListener('error', resolve));
  });

  test('happy path without interims', async () => {
    speechRecognition.start();
    await recognizer.waitForRecognizeOnceAsync();

    // This will fire "firstAudioChunk" on "emitRead"
    recognizer.setFirstAudioChunk();

    recognizer.audioConfig.emitEvent('AudioSourceReadyEvent');

    // cognitiveservices:audioSourceReady
    // webspeech:start
    // webspeech:audiostart

    recognizer.audioConfig.emitRead();

    // cognitiveservices:firstAudioChunk
    // webspeech:soundstart

    recognizer.recognizing(this, MOCK_RECOGNIZING_EVENT);

    // cognitiveservices:recognizing
    // webspeech:speechstart

    recognizer.recognized(this, MOCK_RECOGNIZED_EVENT);

    // cognitiveservices:recognized

    recognizer.audioConfig.emitEvent('AudioSourceOffEvent');

    // cognitiveservices:audioSourceOff
    // webspeech:speechend
    // webspeech:soundend
    // webspeech:audioend

    recognizer.resolveRecognizeOnceAsync(MOCK_SUCCESS_RESULT);

    // cognitiveservices:success
    // webspeech:result.isFinal=true
    // webspeech:end

    await endEventEmitted;

    expect(events).toMatchSnapshot();
  });

  test('happy path with 2 interims', async () => {
    speechRecognition.start();
    speechRecognition.interimResults = true;
    await recognizer.waitForRecognizeOnceAsync();

    // This will fire "firstAudioChunk" on "emitRead"
    recognizer.setFirstAudioChunk();

    recognizer.audioConfig.emitEvent('AudioSourceReadyEvent');

    // cognitiveservices:audioSourceReady
    // webspeech:start
    // webspeech:audiostart

    recognizer.audioConfig.emitRead();

    // cognitiveservices:firstAudioChunk
    // webspeech:soundstart

    recognizer.recognizing(this, MOCK_RECOGNIZING_EVENT);

    // cognitiveservices:recognizing
    // webspeech:speechstart

    recognizer.recognizing(this, MOCK_RECOGNIZING_EVENT);

    // cognitiveservices:recognizing
    // webspeech:result.isFinal=false

    recognizer.recognized(this, MOCK_RECOGNIZED_EVENT);

    // cognitiveservices:recognized

    recognizer.audioConfig.emitEvent('AudioSourceOffEvent');

    // cognitiveservices:audioSourceOff
    // webspeech:speechend
    // webspeech:soundend
    // webspeech:audioend

    recognizer.resolveRecognizeOnceAsync(MOCK_SUCCESS_RESULT);

    // cognitiveservices:success
    // webspeech:result.isFinal=true
    // webspeech:end

    await endEventEmitted;

    expect(events).toMatchSnapshot();
  });

  test('muted microphone', async () => {
    speechRecognition.start();
    await recognizer.waitForRecognizeOnceAsync();

    // This will fire "firstAudioChunk" on "emitRead"
    recognizer.setFirstAudioChunk();

    recognizer.audioConfig.emitEvent('AudioSourceReadyEvent');
    recognizer.audioConfig.emitRead();

    // cognitiveservices:audioSourceReady
    // webspeech:start
    // webspeech:audiostart

    recognizer.recognized(this, {
      offset: 50000000,
      result: {
        duration: 0,
        json: JSON.stringify({
          RecognitionStatus: 'InitialSilenceTimeout',
          Offset: 50000000,
          Duration: 0
        }),
        offset: 50000000,
        reason: 0
      }
    });

    // cognitiveservices:recognized

    recognizer.audioConfig.emitEvent('AudioSourceOffEvent');

    // cognitiveservices:audioSourceOff
    // webspeech:audioend

    recognizer.resolveRecognizeOnceAsync({
      duration: 0,
      json: JSON.stringify({
        RecognitionStatus: 'InitialSilenceTimeout',
        Offset: 50000000,
        Duration: 0
      }),
      offset: 50000000,
      reason: 0
    });

    // cognitiveservices:success
    // webspeech:result.isFinal=true
    // webspeech:end

    await errorEventEmitted;

    expect(events).toMatchSnapshot();
  });

  test('network error', async () => {
    speechRecognition.start();

    await recognizer.waitForRecognizeOnceAsync();

    recognizer.audioConfig.emitEvent('AudioSourceReadyEvent');

    // cognitiveservices:audioSourceReady
    // webspeech:start
    // webspeech:audiostart

    recognizer.canceled(
      this,
      {
        errorDetails: 'Unable to contact server. StatusCode: 1006, Reason: ',
        reason: 0
      }
    );

    // cognitiveservices:canceled
    // audioend
    // error
    // end

    recognizer.rejectRecognizeOnceAsync('Unable to contact server. StatusCode: 1006, Reason: ');

    await errorEventEmitted;

    expect(events).toMatchSnapshot();
  });

  test('microphone blocked', async () => {
    speechRecognition.start();

    await recognizer.waitForRecognizeOnceAsync();

    recognizer.canceled(
      this,
      {
        errorDetails: 'Error occurred during microphone initialization: NotAllowedError: Permission denied',
        reason: 0
      }
    );

    recognizer.rejectRecognizeOnceAsync('Error occurred during microphone initialization: NotAllowedError: Permission denied');

    await errorEventEmitted;

    expect(events).toMatchSnapshot();
  });

  test('abort with partial recognized text', async () => {
    speechRecognition.start();
    await recognizer.waitForRecognizeOnceAsync();

    // This will fire "firstAudioChunk" on "emitRead"
    recognizer.setFirstAudioChunk();

    recognizer.audioConfig.emitEvent('AudioSourceReadyEvent');

    // cognitiveservices:audioSourceReady
    // webspeech:start
    // webspeech:audiostart

    recognizer.audioConfig.emitRead();

    // cognitiveservices:firstAudioChunk
    // webspeech:soundstart

    recognizer.recognizing(this, MOCK_RECOGNIZING_EVENT);

    // cognitiveservices:recognizing
    // webspeech:speechstart

    speechRecognition.abort();

    // cognitiveservices:abort

    recognizer.audioConfig.emitEvent('AudioSourceOffEvent');

    // cognitiveservices:audioSourceOff
    // webspeech:speechend
    // webspeech:soundend
    // webspeech:audioend

    recognizer.rejectRecognizeOnceAsync('aborted');

    // webspeech:error
    // webspeech:end

    await errorEventEmitted;

    expect(events).toMatchSnapshot();
  });
});
