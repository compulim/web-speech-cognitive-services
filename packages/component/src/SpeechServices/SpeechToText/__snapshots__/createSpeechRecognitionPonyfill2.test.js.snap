// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mock SpeechRecognizer with abort after audiostart 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:abort",
  "cognitiveservices:audioSourceOff",
  "webspeech:audioend",
  "webspeech:error { error: 'aborted' }",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with abort after recognizing 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognizing",
  "webspeech:speechstart",
  "cognitiveservices:abort",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "webspeech:error { error: 'aborted' }",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with abort after soundstart 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:abort",
  "cognitiveservices:audioSourceOff",
  "webspeech:soundend",
  "webspeech:audioend",
  "webspeech:error { error: 'aborted' }",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with happy path with 2 interims 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognizing",
  "webspeech:speechstart",
  "webspeech:result { isFinal: false, [[\\"hello\\"]]",
  "cognitiveservices:recognizing",
  "webspeech:result { isFinal: false, [[\\"hello\\"]]",
  "cognitiveservices:recognized",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:result { isFinal: true, [[\\"Hello, John.\\"]]",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with happy path without interims 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognizing",
  "webspeech:speechstart",
  "cognitiveservices:recognized",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:result { isFinal: true, [[\\"Hello, John.\\"]]",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with microphone blocked 1`] = `
Array [
  "cognitiveservices:canceled",
  "webspeech:error { error: 'not-allowed' }",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with muted microphone 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognized",
  "cognitiveservices:audioSourceOff",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:error { error: 'no-speech' }",
  "webspeech:end",
]
`;

exports[`Mock SpeechRecognizer with network error 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:canceled",
  "webspeech:audioend",
  "webspeech:error { error: 'network' }",
  "webspeech:end",
]
`;

exports[`SpeechRecognizer with text normalization of ITN should result in ITN 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognized",
  "webspeech:speechstart",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:result { isFinal: true, [[\\"no (ITN)\\"]]",
  "webspeech:end",
]
`;

exports[`SpeechRecognizer with text normalization of lexical should result in lexical 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognized",
  "webspeech:speechstart",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:result { isFinal: true, [[\\"no (Lexical)\\"]]",
  "webspeech:end",
]
`;

exports[`SpeechRecognizer with text normalization of masked ITN should result in masked ITN 1`] = `
Array [
  "cognitiveservices:audioSourceReady",
  "webspeech:start",
  "webspeech:audiostart",
  "cognitiveservices:firstAudioChunk",
  "webspeech:soundstart",
  "cognitiveservices:recognized",
  "webspeech:speechstart",
  "cognitiveservices:audioSourceOff",
  "webspeech:speechend",
  "webspeech:soundend",
  "webspeech:audioend",
  "cognitiveservices:success",
  "webspeech:result { isFinal: true, [[\\"no (MaskedITN)\\"]]",
  "webspeech:end",
]
`;
